sw/device/lib/dif/dif_usbdev.h

```diff
@@ -0,0 +1,772 @@
+  kDifUsbdevLinkStateDisconnected,
+  kDifUsbdevLinkStatePowered,
+  kDifUsbdevLinkStatePoweredSuspend,
+  kDifUsbdevLinkStateActive,
+  kDifUsbdevLinkStateSuspend,
+} dif_usbdev_link_state_t;
+
+/**
+ * Get USB link state.
+ *
+ * @param usbdev A USB device.
+ * @param link_state USB link state.
+ * @return The result of the operation.
+ */
+dif_usbdev_result_t dif_usbdev_status_get_link_state(
+    dif_usbdev_t *usbdev, dif_usbdev_link_state_t *link_state);
+
+/**
+ * Get the state of the sense pin.
+ *
+ * @param usbdev A USB device.
+ * @param sense State of the sense pin. `true` if the host is providing VBUS,
+ * `false` otherwise.
+ * @return The result of the operation.
+ */
+dif_usbdev_result_t dif_usbdev_status_get_sense(dif_usbdev_t *usbdev,
+                                                bool *sense);
+
+/**
+ * Get the depth of the AV FIFO.
+ *
+ * See also: `dif_usbdev_fill_available_fifo`.
+ *
+ * @param usbdev A USB device.
+ * @param depth Depth of the AV FIFO.
+ * @return The result of the operation.
+ */
+dif_usbdev_result_t dif_usbdev_status_get_available_fifo_depth(
+    dif_usbdev_t *usbdev, uint8_t *depth);
+/**
+ * Check if AV FIFO is full.
+ *
+ * See also: `dif_usbdev_fill_available_fifo`.
+ *
+ * @param usbdev A USB device.
+ * @param is_full State of the AV FIFO. `true` if full, false otherwise.
+ * @return The result of the operation.
+ */
+dif_usbdev_result_t dif_usbdev_status_get_available_fifo_full(
+    dif_usbdev_t *usbdev, bool *is_full);
+/**
+ * Get the depth of the RX FIFO.
+ *
+ * See also: `dif_usbdev_rx_packet_get_info`.
+ *
+ * @param usbdev A USB device.
+ * @param depth Depth of the RX FIFO.
+ * @return The result of the operation.
+ */
+dif_usbdev_result_t dif_usbdev_status_get_rx_fifo_depth(dif_usbdev_t *usbdev,
+                                                        uint8_t *depth);
+
+/**
+ * Check if the RX FIFO is empty.
+ *
+ * See also: `dif_usbdev_rx_packet_get_info`.
+ *
+ * @param usbdev A USB device.
+ * @param is_empty State of the RX FIFO. `true` if empty, `false` otherwise.
+ * @return The result of the operation.
+ */
+dif_usbdev_result_t dif_usbdev_status_get_rx_fifo_empty(dif_usbdev_t *usbdev,
+                                                        bool *is_empty);
+
+/**
+ * USB device interrupts.
+ *
+ * `kDifUsbdevIrqFirst` and `kDifUsbdevIrqLast` are provided to let clients
+ * iterate over all available interrupts.
+ */
+typedef enum dif_usbdev_irq {
+  kDifUsbdevIrqFirst,
+  kDifUsbdevIrqPktReceived = kDifUsbdevIrqFirst,
+  kDifUsbdevIrqPktSent,
+  kDifUsbdevIrqDisconnected,
+  kDifUsbdevIrqHostLost,
+  kDifUsbdevIrqLinkReset,
+  kDifUsbdevIrqLinkSuspend,
+  kDifUsbdevIrqLinkResume,
+  kDifUsbdevIrqAvEmpty,
+  kDifUsbdevIrqRxFull,
+  kDifUsbdevIrqAvOverflow,
+  kDifUsbdevIrqLinkInError,
+  kDifUsbdevIrqRxCrcError,
+  kDifUsbdevIrqRxPidError,
+  kDifUsbdevIrqRxBitstuffError,
+  kDifUsbdevIrqFrame,
+  kDifUsbdevIrqConnected,
+  kDifUsbdevIrqLast = kDifUsbdevIrqConnected
```

Comment:
Other DIFs currently seem to mark `kLast` and friends as `\internal` in the doc comment. Consider doing this and removing the reference to them in this enum's overall type comment.